src/masternodes/mn_checks.cpp

```diff
@@ -1259,15 +1269,27 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
 
             srcAddr = CScript(order->creationTx.begin(), order->creationTx.end());
 
+            CScript offerTxidAddr(offer->creationTx.begin(), offer->creationTx.end());
+
             CAmount calcAmount(static_cast<CAmount>((arith_uint256(submitdfchtlc.amount) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
             if (calcAmount > offer->amount)
                 return Res::Err("amount must be lower or equal the offer one");
 
-            CScript offerTxidAddr(offer->creationTx.begin(), offer->creationTx.end());
-
-            //calculating adjusted takerFee
-            CAmount BTCAmount(static_cast<CAmount>((arith_uint256(submitdfchtlc.amount) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
-            auto takerFee = CalculateTakerFee(BTCAmount);
+            CAmount takerFee = offer->takerFee;
+            //EunosPaya: calculating adjusted takerFee only if amount in htlc different than in offer
+            if (static_cast<int>(height) >= consensus.EunosPayaHeight)
+            {
+                if (calcAmount < offer->amount)
+                {
+                    CAmount BTCAmount(static_cast<CAmount>((arith_uint256(submitdfchtlc.amount) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
+                    takerFee = static_cast<CAmount>((arith_uint256(BTCAmount) * arith_uint256(offer->takerFee) / arith_uint256(offer->amount)).GetLow64());
+                }
+            }
+            else
+            {
+                CAmount BTCAmount(static_cast<CAmount>((arith_uint256(submitdfchtlc.amount) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
+                takerFee = CalculateTakerFee(BTCAmount);
+            }
 
             // refund the rest of locked takerFee if there is difference
             if (offer->takerFee - takerFee) {
```

Comment:
If this value is negative then still have problem

Reply:
It will go into else before EunosPaya, and after it will enter the if and calculate the takerFee as ratio of the current one so `offer->takerFee - takerFee` will never be negative.